prow/gcs/reporter/reporter.go

```diff
@@ -0,0 +1,228 @@
+	}
+	output, err := json.Marshal(f)
+	if err != nil {
+		return err
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return err
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload started.json to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "started.json")).If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+// reportFinishedJob uploads a finished.json for the job, iff one did not already exist.
+func (gr *gcsReporter) reportFinishedJob(ctx context.Context, pj *prowv1.ProwJob) error {
+	// We don't upload a finished.json until we finish.
+	if !pj.Complete() {
+		return nil
+	}
+
+	completion := pj.Status.CompletionTime.Unix()
+	passed := pj.Status.State == prowv1.SuccessState
+	f := metadata.Finished{
+		Timestamp: &completion,
+		Passed:    &passed,
+		Metadata:  metadata.Metadata{"uploader": "crier"},
+		Result:    string(pj.Status.State),
+	}
+	output, err := json.Marshal(f)
+	if err != nil {
+		return err
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return err
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload finished.json info to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "finished.json")).If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+func (gr *gcsReporter) reportProwjob(ctx context.Context, pj *prowv1.ProwJob) error {
+	// Unconditionally dump the prowjob to GCS, on all job updates.
+	output, err := json.Marshal(pj)
+	if err != nil {
+		return err
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return err
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload pod info to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "prowjob.json")).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+func (gr *gcsReporter) writeContent(w *storage.Writer, content []byte) error {
+	_, err := w.Write(content)
+	var reportErr error
+	if isErrUnexpected(err) {
+		reportErr = err
+		gr.logger.WithError(err).WithFields(logrus.Fields{"bucket": w.Bucket, "name": w.Name}).Warn("Uploading info to GCS failed (write)")
+	}
+	err = w.Close()
+	if isErrUnexpected(err) {
+		reportErr = err
+		gr.logger.WithError(err).WithFields(logrus.Fields{"bucket": w.Bucket, "name": w.Name}).Warn("Uploading info to GCS failed (close)")
+	}
+	return reportErr
+}
+
+func isErrUnexpected(err error) bool {
+	if err == nil {
+		return false
+	}
+	// Precondition Failed is expected and we can silently ignore it.
+	if e, ok := err.(*googleapi.Error); ok {
+		if e.Code == http.StatusPreconditionFailed {
+			return false
+		}
+		return true
+	}
+	return true
+}
+
+func (gr *gcsReporter) getJobDestination(pj *prowv1.ProwJob) (bucket string, dir string, err error) {
```

Comment:
Can you use gcs.Path instead of two strings?

https://github.com/GoogleCloudPlatform/testgrid/blob/master/util/gcs/gcs.go

(and `bucket, dir string`)

Reply:
I can, but I'm not sure it's worth it - I'd have to construct a string so that `gcs.Path` can parse it back (which can fail). The increased complexity / extra package doesn't seem worth the benefit of returning one thing instead of two.